### DISCUSSION TITLE
Weekly Meeting - 22nd Feb 2022

### DISCUSSION THREAD

**Discussion by AdamFrisby** [omigroup/omi-scripting-group#22] [General]
Discussion while meeting is running for sharing notes

**Comment by robertlong** [comment]
Today I shared a presentation of the API I'm using in my demo: https://github.com/robertlong/omi-wasm-group-experiment

I'm still working on the implementation, but the work is slow with other work/life things lately.

Some of the work is happening in this repo: https://github.com/Web-ECS/do-three

Nate has been working on bridging Three.js and ECS concepts in a way that is distilled down to data oriented design concepts. I thought it fit really well with the performance bottlenecks we face today in WebAssembly. In that most things are limited by the speed of function calls across the WASM boundary (something to validate in future meetings).

Questions/Topics from the presentation:
- This low level API could be wrapped in a MonoBehaviour or Three.js Object3D.
- How might transforms work when performing bulk operations. Do we do something like transactions or a dirty flag and an update function? What about techniques such as Read/Write modifiers on queries in ECS libraries such as Specs?
- Is there a need for both a `prev_sibling` and `next_sibling` property? I pointed people to the prior work for that https://bitsquid.blogspot.com/2014/10/building-data-oriented-entity-system.html
- Can property stores be dynamically allocated, resized, etc? Is it beneficial to have a fixed size and allocate up front, especially when defining limits for a sandbox?
- We drew similarities between this API and DOM APIs. What will people want to build on top of this API? Should we design for people building complex systems iterating over lots of entities? React/[fastdom](https://github.com/wilsonpage/fastdom) as examples of devs trying to improve DOM performance through virtualization/batching.
- We talked about permissions. What permissions will this script have? Which objects can it access in the scene?

Next week we'll dive deeper into how read/write operations can be accomplished in the ABI. We'll focus on performance and interoperability with existing libraries/techniques. As well as real world scenarios we should design around. @AdamFrisby said he may be able to do some benchmarking of various operations before then. I may be able to as well.
